src/main/java/io/reactivex/internal/operators/parallel/ParallelJoin.java

```
@@ -0,0 +1,359 @@
+
+        final AtomicReference<Throwable> error = new AtomicReference<Throwable>();
+
+        final AtomicLong requested = new AtomicLong();
+
+        volatile boolean cancelled;
+
+        final AtomicInteger done = new AtomicInteger();
+
+        JoinSubscription(Subscriber<? super T> actual, int n, int prefetch) {
+            this.actual = actual;
+            @SuppressWarnings("unchecked")
+            JoinInnerSubscriber<T>[] a = new JoinInnerSubscriber[n];
+
+            for (int i = 0; i < n; i++) {
+                a[i] = new JoinInnerSubscriber<T>(this, prefetch);
+            }
+
+            this.subscribers = a;
+            done.lazySet(n);
+        }
+
+        @Override
+        public void request(long n) {
+            if (SubscriptionHelper.validate(n)) {
+                BackpressureHelper.add(requested, n);
+                drain();
+            }
+        }
+
+        @Override
+        public void cancel() {
+            if (!cancelled) {
+                cancelled = true;
+
+                cancelAll();
+
+                if (getAndIncrement() == 0) {
+                    cleanup();
+                }
+            }
+        }
+
+        void cancelAll() {
+            for (JoinInnerSubscriber<T> s : subscribers) {
+                s.cancel();
+            }
+        }
+
+        void cleanup() {
+            for (JoinInnerSubscriber<T> s : subscribers) {
+                s.queue = null;
+            }
+        }
+
+        void onNext(JoinInnerSubscriber<T> inner, T value) {
+            if (get() == 0 && compareAndSet(0, 1)) {
+                if (requested.get() != 0) {
+                    actual.onNext(value);
+                    if (requested.get() != Long.MAX_VALUE) {
+                        requested.decrementAndGet();
+                    }
+                    inner.request(1);
+                } else {
+                    SimpleQueue<T> q = inner.getQueue();
+
+                    // FIXME overflow handling
```

Signal MBE? When do you plan to resolve FIXME? It'll lead to silently dropped values…

---

Adding it right now.